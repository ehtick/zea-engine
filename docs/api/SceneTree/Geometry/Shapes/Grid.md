<a name="Grid"></a>

### Grid 
Represents a network of lines that cross each other to form a series of squares or rectangles.

```
const grid = new Grid(5, 5, 50, 50, true)
```

**Parameters**
* **X([`NumberParameter`](api/SceneTree\Parameters\NumberParameter.md)):** Length of the grid along the `X` axis.
* **Y([`NumberParameter`](api/SceneTree\Parameters\NumberParameter.md)):** Length of the grid along the `Y` axis.
* **XDivisions([`NumberParameter`](api/SceneTree\Parameters\NumberParameter.md)):** Number of divisions along `X` axis
* **YDivisions([`NumberParameter`](api/SceneTree\Parameters\NumberParameter.md)):** Number of divisions along `Y` axis
* **SkipCenterLines([`BooleanParameter`](api/SceneTree\Parameters\BooleanParameter.md)):** Property that indicates whether to display the center grid lines or not


**Extends**: <code>[ProceduralLines](api/SceneTree\Geometry\Shapes\ProceduralLines.md)</code>  
<a name="new_Grid_new"></a>

### new Grid
Create a grid.


| Param | Type | Default | Description |
| --- | --- | --- | --- |
| x | <code>number</code> | <code>1</code> | The length of the grid along the `X` axis. |
| y | <code>number</code> | <code>1</code> | The length of the grid along the `Y` axis. |
| xDivisions | <code>number</code> | <code>10</code> | The number of divisions along `X` axis. |
| yDivisions | <code>number</code> | <code>10</code> | The number of divisions along `Y` axis. |
| skipCenterLines | <code>boolean</code> | <code>false</code> | A boolean indicating whether to display the center grid lines or not. |



### [Class Tests](api/SceneTree\Geometry\Shapes/Grid.test)