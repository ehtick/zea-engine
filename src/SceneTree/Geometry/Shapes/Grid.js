import { Vec2 } from '../../../Math/Vec2';
import { Vec3 } from '../../../Math/Vec3';
import { Lines } from '../Lines.js';

class Grid extends Lines {
  constructor(x = 1.0, y = 1.0, xDivisions = 10, yDivisions = 10, skipCenterLines=false) {
    super();

    if(isNaN(x) || isNaN(y) || isNaN(xDivisions) || isNaN(yDivisions))
      throw("Invalid geom args");
    
    this.__x = x;
    this.__y = y;
    this.__xDivisions = xDivisions;
    this.__yDivisions = yDivisions;
    this.__skipCenterLines = skipCenterLines;
    this.__rebuild();
  }

  get sizeX() {
    return this.__x;
  }

  set sizeX(val) {
    this.__x = val;
    this.__resize();
  }

  get sizeY() {
    return this.__y;
  }

  set sizeY(val) {
    this.__y = val;
    this.__resize();
  }

  get divisionsX() {
    return this.__xDivisions;
  }

  set divisionsX(val) {
    this.__xDivisions = val;
    this.__rebuild();
  }

  get divisionsY() {
    return this.__yDivisions;
  }

  set divisionsY(val) {
    this.__yDivisions = val;
    this.__rebuild();
  }

  setSize(x, z) {
    this.__x = x;
    this.__y = z;
    this.__resize();
  }

  __rebuild() {
    const skipCenterLines = this.__skipCenterLines && (this.__xDivisions % 2) == 0 && (this.__yDivisions % 2) == 0;
    this.setNumVertices((this.__xDivisions + this.__yDivisions + 2 - (skipCenterLines ? 1 : 0)) * 2);
    this.setNumSegments(this.__xDivisions + this.__yDivisions + 2 - (skipCenterLines ? 1 : 0));
    let idx = 0;
    for (let i = 0; i <= this.__xDivisions; i++) {
      if(skipCenterLines && i == this.__xDivisions/2)
        continue;
      let v0 = (idx*2);
      let v1 = ((idx*2) + 1);
      this.setSegment(idx, v0, v1);
      idx++;
    }
    for (let i=0; i <= this.__yDivisions; i++) {
      if(skipCenterLines && i == this.__xDivisions/2)
        continue;
      let v0 = (idx*2);
      let v1 = ((idx*2) + 1);
      this.setSegment(idx, v0, v1);
      idx++;
    }
    this.__resize();
  }

  __resize() {
    const skipCenterLines = this.__skipCenterLines && (this.__xDivisions % 2) == 0 && (this.__yDivisions % 2) == 0;
    let idx = 0;
    for (let i = 0; i <= this.__xDivisions; i++) {
      if(skipCenterLines && i == this.__xDivisions/2)
        continue;
      let v0 = (idx*2);
      let v1 = ((idx*2) + 1);
      let x = ((i / (this.__xDivisions) - 0.5)) * this.__x;
      this.getVertex(v0).set(x, -0.5 * this.__y, 0.0);
      this.getVertex(v1).set(x,  0.5 * this.__y, 0.0);
      idx++;
    }
    for (let i = 0; i <= this.__yDivisions; i++) {
      if(skipCenterLines && i == this.__xDivisions/2)
        continue;
      let v0 = (idx*2);
      let v1 = ((idx*2) + 1);
      let y = ((i / (this.__yDivisions) - 0.5)) * this.__y;
      this.getVertex(v0).set(-0.5 * this.__x, y, 0.0);
      this.getVertex(v1).set( 0.5 * this.__x, y, 0.0);
      idx++;
    }

    this.setBoundingBoxDirty();
  }

  toJSON() {
    let json = super.toJSON();
    json['x'] = this.__x;
    json['z'] = this.__y;
    json['xDivisions'] = this.__xDivisions;
    json['yDivisions'] = this.__yDivisions;
    return json
  }
};

export {
  Grid
};
//export default Grid;