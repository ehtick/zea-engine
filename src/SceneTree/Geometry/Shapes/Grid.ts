import { BooleanParameter, NumberParameter } from '../../Parameters/index'
import { Registry } from '../../../Registry'
import { ProceduralLines } from './ProceduralLines'
import { Vec3Attribute } from '../Vec3Attribute'

/**
 * Represents a network of lines that cross each other to form a series of squares or rectangles.
 *
 * ```
 * const grid = new Grid(5, 5, 50, 50, true)
 * ```
 *
 * **Parameters**
 * * **X(`NumberParameter`):** Length of the grid along the `X` axis.
 * * **Y(`NumberParameter`):** Length of the grid along the `Y` axis.
 * * **XDivisions(`NumberParameter`):** Number of divisions along `X` axis
 * * **YDivisions(`NumberParameter`):** Number of divisions along `Y` axis
 * * **SkipCenterLines(`BooleanParameter`):** Property that indicates whether to display the center grid lines or not
 *
 * @extends {ProceduralLines}
 */
class Grid extends ProceduralLines {
  /**
   * @member skipCenterLinesParam - Property that indicates whether to display the center grid lines or not
   */
  skipCenterLinesParam: BooleanParameter

  /**
   * @member xDivisionsParam - Number of divisions along `X` axis
   */
  xDivisionsParam: NumberParameter

  /**
   * @member xParam - Length of the grid along the `X` axis.
   */
  xParam: NumberParameter

  /**
   * @member yDivisionsParam - Number of divisions along `Y` axis
   */
  yDivisionsParam: NumberParameter

  /**
   * @member yParam - Length of the grid along the `Y` axis.
   */
  yParam: NumberParameter
  // topologyParams: string[]

  /**
   * Create a grid.
   * @param x - The length of the grid along the `X` axis.
   * @param y - The length of the grid along the `Y` axis.
   * @param xDivisions - The number of divisions along `X` axis.
   * @param yDivisions - The number of divisions along `Y` axis.
   * @param skipCenterLines - A boolean indicating whether to display the center grid lines or not.
   */
  constructor(x = 1.0, y = 1.0, xDivisions = 10, yDivisions = 10, skipCenterLines = false) {
    super()
    this.topologyParams = []
    if (isNaN(x) || isNaN(y) || isNaN(xDivisions) || isNaN(yDivisions)) throw new Error('Invalid geom args')

    this.xParam = this.addParameter(new NumberParameter('X', x)) as NumberParameter
    this.yParam = this.addParameter(new NumberParameter('Y', y)) as NumberParameter
    this.xDivisionsParam = this.addParameter(new NumberParameter('XDivisions', xDivisions)) as NumberParameter
    this.yDivisionsParam = this.addParameter(new NumberParameter('YDivisions', yDivisions)) as NumberParameter
    this.skipCenterLinesParam = this.addParameter(
      new BooleanParameter('SkipCenterLines', skipCenterLines)
    ) as BooleanParameter

    this.topologyParams.push('XDivisions')
    this.topologyParams.push('YDivisions')
    this.topologyParams.push('SkipCenterLines')
  }

  /**
   * The rebuild method.
   * @private
   */
  rebuild(): void {
    const xDivisions = this.xDivisionsParam.value || 10
    const yDivisions = this.yDivisionsParam.value || 10

    const skipCenterLines = this.skipCenterLinesParam.value && xDivisions % 2 == 0 && yDivisions % 2 == 0
    this.setNumVertices((xDivisions + yDivisions + 2 - (skipCenterLines ? 1 : 0)) * 2)
    this.setNumSegments(xDivisions + yDivisions + 2 - (skipCenterLines ? 1 : 0))
    let idx = 0
    for (let i = 0; i <= xDivisions; i++) {
      if (skipCenterLines && i == xDivisions / 2) continue
      const v0 = idx * 2
      const v1 = idx * 2 + 1
      this.setSegmentVertexIndices(idx, v0, v1)
      idx++
    }
    for (let i = 0; i <= yDivisions; i++) {
      if (skipCenterLines && i == xDivisions / 2) continue
      const v0 = idx * 2
      const v1 = idx * 2 + 1
      this.setSegmentVertexIndices(idx, v0, v1)
      idx++
    }
    this.resize()
  }

  /**
   * The resize method.
   * @private
   */
  resize(): void {
    const positions = <Vec3Attribute>this.getVertexAttribute('positions')
    const xDivisions = this.xDivisionsParam.value || 10
    const yDivisions = this.yDivisionsParam.value || 10
    const xSize = this.xParam.value || 1.0
    const ySize = this.yParam.value || 1.0

    const skipCenterLines = this.skipCenterLinesParam.value && xDivisions % 2 == 0 && yDivisions % 2 == 0
    let idx = 0
    for (let i = 0; i <= xDivisions; i++) {
      if (skipCenterLines && i == xDivisions / 2) continue
      const v0 = idx * 2
      const v1 = idx * 2 + 1
      const x = (i / xDivisions - 0.5) * xSize
      if (positions) {
        positions.getValueRef(v0).set(x, -0.5 * ySize, 0.0)
        positions.getValueRef(v1).set(x, 0.5 * ySize, 0.0)
      }

      idx++
    }
    for (let i = 0; i <= yDivisions; i++) {
      if (skipCenterLines && i == xDivisions / 2) continue
      const v0 = idx * 2
      const v1 = idx * 2 + 1
      const y = (i / yDivisions - 0.5) * ySize
      if (positions) {
        positions.getValueRef(v0).set(-0.5 * xSize, y, 0.0)
        positions.getValueRef(v1).set(0.5 * xSize, y, 0.0)
      }
      idx++
    }
  }
}

Registry.register('Grid', Grid)

export { Grid }
