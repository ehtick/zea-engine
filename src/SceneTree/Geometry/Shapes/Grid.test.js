import { Grid } from './Grid'

describe('Grid', () => {
  it('tests default parameters', () => {
    const grid = new Grid()

    expect(grid.getParameter('X').getValue()).toBe(1)
    expect(grid.getParameter('Y').getValue()).toBe(1)
    expect(grid.getParameter('XDivisions').getValue()).toBe(10)
    expect(grid.getParameter('YDivisions').getValue()).toBe(10)
    expect(grid.getParameter('SkipCenterLines').getValue()).toBe(false)
  })

  it('updates parameters', () => {
    const grid = new Grid()
    grid.getParameter('X').setValue(2)
    grid.getParameter('Y').setValue(2)
    grid.getParameter('XDivisions').setValue(20)
    grid.getParameter('YDivisions').setValue(15)
    grid.getParameter('SkipCenterLines').setValue(true)

    expect(grid.getParameter('X').getValue()).toBe(2)
    expect(grid.getParameter('Y').getValue()).toBe(2)
    expect(grid.getParameter('XDivisions').getValue()).toBe(20)
    expect(grid.getParameter('YDivisions').getValue()).toBe(15)
    expect(grid.getParameter('SkipCenterLines').getValue()).toBe(true)
  })

  // There's an issue with flags, that are preventing parameters to be exported.
  it('saves to JSON (serialization).', () => {
    const grid = new Grid(4, 4, 2, 2, true)
    const outputJSON = grid.toJSON()

    expect(outputJSON).toMatchSnapshot()
  })

  it('restores from JSON (serialization).', () => {
    const grid = new Grid()
    const inputJSON = {
      params: {
        X: { value: 4 },
        Y: { value: 4 },
        XDivisions: { value: 2 },
        YDivisions: { value: 2 },
        SkipCenterLines: { value: true },
      },
      type: 'Grid',
      vertexAttributes: {},
      indices: [0, 1, 2, 3, 4, 5, 6, 7, 0, 0],
    }
    grid.fromJSON(inputJSON)

    const newGrid = new Grid(4, 4, 2, 2, true)
    expect(grid.toJSON()).toEqual(newGrid.toJSON())
  })
})
