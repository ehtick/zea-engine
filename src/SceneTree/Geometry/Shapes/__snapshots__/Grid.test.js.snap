// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid saves to JSON (serialization). 1`] = `
Object {
  "params": Object {
    "SkipCenterLines": Object {
      "value": true,
    },
    "X": Object {
      "value": 4,
    },
    "XDivisions": Object {
      "value": 2,
    },
    "Y": Object {
      "value": 4,
    },
    "YDivisions": Object {
      "value": 2,
    },
  },
  "type": "Grid",
  "vertexAttributes": Object {},
}
`;
